File: pydss/conversion/EquivalentObjImpl.py
from dss.conversion.EquivalentObj import (EquivalentObj,)
from dss.common.impl.Utilities import (Utilities,)
from dss.common.impl.DSSGlobals import (DSSGlobals,)
from dss.conversion.impl.PCElementImpl import (PCElementImpl,)
from dss.parser.impl.Parser import (Parser,)
from dss.conversion.Equivalent import (Equivalent,)
from dss.shared.impl.ComplexUtil import (ComplexUtil,)
from dss.shared.impl.CMatrixImpl import (CMatrixImpl,)
# from java.io.PrintStream import (PrintStream,)
# from org.apache.commons.math.complex.Complex import (Complex,)


class EquivalentObjImpl(PCElementImpl, EquivalentObj):
    kVBase = None
    VMag = None
    perUnit = None
    angle = None
    equivFrequency = None
    R1 = None
    X1 = None
    R0 = None
    X0 = None
    needToDoRecalc = None
    # base frequency series Z matrix
    Z = None
    ZInv = None

    def __init__(self, parClass, sourceName):
        super(EquivalentObjImpl, self)(parClass)
        self.setName(sourceName.toLowerCase())
        self.objType = parClass.getDSSClassType()
        # SOURCE + NON_PCPD_ELEM;  // don't want this in PC element list
        self.setNPhases(3)
        self.nConds = 3
        self.setNTerms(1)
        self.Z = None
        self.ZInv = None
        # Basefrequency = 60.0;
        # set in base class
        self.R1 = None
        self.X1 = None
        self.R0 = None
        self.X0 = None
        self.reallocRX()
        self.R1[0] = 1.65
        self.X1[0] = 6.6
        self.R0[0] = 1.9
        self.X0[0] = 5.7
        self.kVBase = 115.0
        self.perUnit = 1.0
        self.equivFrequency = self.baseFrequency
        self.angle = 0.0
        self.spectrum = 'defaultvsource'
        self.initPropertyValues(0)
        self.YOrder = self.nTerms * self.nConds
        self.recalcElementData()

    def idx(self, a, b):
        return ((b - 1) * self.nTerms) + a

    def recalcElementData(self):
        # int i, j, ii, jj;
        if self.Z is not None:
            self.Z = None
        if self.ZInv is not None:
            self.ZInv = None
            # for a source, nPhases = nCond, for now
        self.Z = CMatrixImpl(self.nPhases * self.nTerms)
        self.ZInv = CMatrixImpl(self.nPhases * self.nTerms)
        # build Z matrix for all phases
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.nTerms):
                break
            _1 = True
            j = 0
            while True:
                if _1 is True:
                    _1 = False
                else:
                    j += 1
                if not (j < self.nTerms):
                    break
                indx = self.idx(i, j)
                Zs = ComplexUtil.divide(Complex((2.0 * self.R1[indx]) + self.R0[indx], (2.0 * self.X1[indx]) + self.X0[indx]), 3.0)
                Zm = ComplexUtil.divide(Complex(self.R0[indx] - self.R1[indx], self.X0[indx] - self.X1[indx]), 3.0)
                iOffset = (i - 1) * self.nPhases
                jOffset = (j - 1) * self.nPhases
                _2 = True
                ii = 0
                while True:
                    if _2 is True:
                        _2 = False
                    else:
                        ii += 1
                    if not (ii < self.nPhases):
                        break
                    _3 = True
                    jj = 0
                    while True:
                        if _3 is True:
                            _3 = False
                        else:
                            jj += 1
                        if not (jj < ii):
                            break
                        if ii == jj:
                            self.Z.setElement(ii + iOffset, jj + jOffset, Zs)
                        else:
                            self.Z.setElement(ii + iOffset, jj + jOffset, Zm)
                            self.Z.setElement(jj + iOffset, ii + jOffset, Zm)
                            # set other off-diagonal in this submatrix
            # voltage source properties
        _4 = self.nPhases
        _5 = False
        while True:
            if _4 == 1:
                _5 = True
                self.VMag = self.kVBase * self.perUnit * 1000.0
                break
            if True:
                _5 = True
                self.VMag = (self.kVBase * self.perUnit * 1000.0) / 2.0 / self.Math.sin(((180.0 / self.nPhases) * self.Math.PI) / 180.0)
                break
            break
        self.setSpectrumObj(DSSGlobals.spectrumClass.find(self.getSpectrum()))
        if self.getSpectrumObj() is None:
            DSSGlobals.doSimpleMsg('Spectrum object \"' + self.getSpectrum() + '\" for device equivalent.' + self.getName() + ' not found.', 802)
        self.setInjCurrent(Utilities.resizeArray(self.getInjCurrent(), self.YOrder))
        self.needToDoRecalc = False

    def calcYPrim(self):
        # build only YPrim series
        if self.isYprimInvalid():
            if self.YPrimSeries is not None:
                self.YPrimSeries = None
            self.YPrimSeries = CMatrixImpl(self.YOrder)
            if self.YPrim is not None:
                self.YPrim = None
            self.YPrim = CMatrixImpl(self.YOrder)
        else:
            self.YPrimSeries.clear()
            self.YPrim.clear()
        if self.needToDoRecalc:
            self.recalcElementData()
        self.YPrimFreq = DSSGlobals.activeCircuit.getSolution().getFrequency()
        freqMultiplier = self.YPrimFreq / self.baseFrequency
        # Put in Series RL matrix adjusted for frequency
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.YOrder):
                break
            _1 = True
            j = 0
            while True:
                if _1 is True:
                    _1 = False
                else:
                    j += 1
                if not (j < self.YOrder):
                    break
                value = self.Z.getElement(i, j)
                # Modify from base freq
                value = Complex(value.getReal(), value.getImaginary() * freqMultiplier)
                self.ZInv.setElement(i, j, value)
        self.ZInv.invert()
        # Invert in place
        if self.ZInv.getErrorCode() > 0:
            # If error, put in large series conductance
            DSSGlobals.doErrorMsg('EquivalentObj.calcYPrim', 'Matrix inversion error for equivalent \"' + self.getName() + '\"', 'Invalid impedance specified. Replaced with small resistance.', 803)
            self.ZInv.clear()
            _2 = True
            i = 0
            while True:
                if _2 is True:
                    _2 = False
                else:
                    i += 1
                if not (i < self.nPhases):
                    break
                self.ZInv.setElement(i, i, Complex(1.0 / DSSGlobals.EPSILON, 0.0))
        self.YPrimSeries.copyFrom(self.ZInv)
        self.YPrim.copyFrom(self.YPrimSeries)
        # Now account for open conductors
        # For any conductor that is open, zero out row and column
        super(EquivalentObjImpl, self).calcYPrim()
        self.setYPrimInvalid(False)

    def getVterminalForSource(self):
        # This formulation will theoretically handle voltage sources of any number
        # of phases assuming they are equally displaced in time.

        try:
            _0 = self.nPhases
            _1 = False
            while True:
                if _0 == 1:
                    _1 = True
                    self.VMag = self.kVBase * self.perUnit * 1000.0
                    break
                if True:
                    _1 = True
                    self.VMag = (self.kVBase * self.perUnit * 1000.0) / 2.0 / self.Math.sin(((180.0 / self.nPhases) * self.Math.PI) / 180.0)
                    break
                break
            sol = DSSGlobals.activeCircuit.getSolution()
            if sol.isHarmonicModel():
                equivHarm = sol.getFrequency() / self.equivFrequency
                VHarm = self.getSpectrumObj().getMult(equivHarm).multiply(self.VMag)
                # base voltage for this harmonic
                VHarm = Utilities.rotatePhasorDeg(VHarm, equivHarm, self.angle)
                # rotate for phase 1 shift
                _2 = True
                i = 0
                while True:
                    if _2 is True:
                        _2 = False
                    else:
                        i += 1
                    if not (i < self.nPhases):
                        break
                    self.VTerminal[i] = VHarm
                    if i < self.nPhases:
                        VHarm = Utilities.rotatePhasorDeg(VHarm, equivHarm, -360.0 / self.nPhases)
            else:
                _3 = True
                i = 0
                while True:
                    if _3 is True:
                        _3 = False
                    else:
                        i += 1
                    if not (i < self.nPhases):
                        break
                    self.VTerminal[i] = ComplexUtil.polarDeg2Complex(self.VMag, (360.0 + self.angle) - (((i - 1) * 360.0) / self.nPhases))
        except Exception, e:
            DSSGlobals.doSimpleMsg('Error computing voltages for Equivalent.' + self.getName() + '. Check specification. Aborting.', 804)
            if DSSGlobals.inRedirect:
                DSSGlobals.redirectAbort = True

    def injCurrents(self):
        self.getInjCurrents(self.getInjCurrent())
        return super(EquivalentObjImpl, self).injCurrents()
        # add into system array

    def getCurrents(self, curr):
        try:
            sol = DSSGlobals.activeCircuit.getSolution()
            _0 = True
            i = 0
            while True:
                if _0 is True:
                    _0 = False
                else:
                    i += 1
                if not (i < self.YOrder):
                    break
                self.VTerminal[i] = sol.getNodeV()[self.nodeRef[i]]
            self.YPrim.vMult(curr, self.VTerminal)
            self.getInjCurrents(self.complexBuffer)
            # get present value of inj currents
            # add together with Yprim currents
            _1 = True
            i = 0
            while True:
                if _1 is True:
                    _1 = False
                else:
                    i += 1
                if not (i < self.YOrder):
                    break
                curr[i] = curr[i].subtract(self.complexBuffer[i])
        except Exception, e:
            DSSGlobals.doErrorMsg('GetCurrents for element: ' + self.getName() + '.', e.getMessage(), 'Inadequate storage allotted for circuit element.', 805)

    def getInjCurrents(self, curr):
        self.getVterminalForSource()
        self.YPrim.vMult(curr, self.VTerminal)
        # I = Y V
        self.setITerminalUpdated(False)

    def dumpProperties(self, f, complete):
        super(EquivalentObjImpl, self).dumpProperties(f, complete)
        pc = self.getParentClass()
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < pc.getNumProperties()):
                break
            print '~ ' + pc.getPropertyName()[i] + '=' + self.getPropertyValue(i)
        if complete:
            print 
            print 'baseFrequency=' + self.baseFrequency
            print 'vMag=' + self.VMag
            print 'zMatrix='
            _1 = True
            i = 0
            while True:
                if _1 is True:
                    _1 = False
                else:
                    i += 1
                if not (i < self.nPhases):
                    break
                _2 = True
                j = 0
                while True:
                    if _2 is True:
                        _2 = False
                    else:
                        j += 1
                    if not (j < i):
                        break
                    c = self.Z.getElement(i, j)
                    f.print_(c.getReal() + ' + j' + c.getImaginary())
                print 

    def initPropertyValues(self, arrayOffset):
        self.propertyValue[0] = '1'
        self.propertyValue[1] = self.getBus(0)
        self.propertyValue[2] = '115'
        self.propertyValue[3] = '1'
        self.propertyValue[4] = '0'
        self.propertyValue[5] = '60'
        self.propertyValue[6] = '3'
        self.propertyValue[7] = '1.65'
        self.propertyValue[8] = '6.6'
        self.propertyValue[9] = '1.9'
        self.propertyValue[10] = '5.7'
        super(EquivalentObjImpl, self).initPropertyValues(Equivalent.NumPropsThisClass)

    def getPropertyValue(self, index):
        _0 = index
        _1 = False
        while True:
            if _0 == 0:
                _1 = True
                return None
            if True:
                _1 = True
                return super(EquivalentObjImpl, self).getPropertyValue(index)
            break

    def makePosSequence(self):
        """Make a positive sequence model."""
        # Private method in OpenDSS
        s = 'Phases=1 '
        s = s + String.format.format('basekV=%-.5g ', self.kVBase / DSSGlobals.SQRT3)
        s = s + String.format.format('r1=%-.5g ', self.R1)
        s = s + String.format.format('x1=%-.5g ', self.X1)
        Parser.getInstance().setCmdString(s)
        self.edit()
        super(EquivalentObjImpl, self).makePosSequence()

    def doTerminalsDef(self, n):
        # Private method in OpenDSS
        result = self.nTerms
        if n != self.nTerms:
            if n > 0:
                self.reallocRX()
        return result

    def parseDblMatrix(self, mat):
        """Parse input string as an array."""
        Parser.getInstance().parseAsSymMatrix(self.nTerms, mat)

    def reallocRX(self):
        self.R1 = Utilities.resizeArray(self.R1, self.Math.pow(self.nTerms, 2))
        self.X1 = Utilities.resizeArray(self.X1, self.Math.pow(self.nTerms, 2))
        self.R0 = Utilities.resizeArray(self.R0, self.Math.pow(self.nTerms, 2))
        self.X0 = Utilities.resizeArray(self.X0, self.Math.pow(self.nTerms, 2))

    def getZ(self):
        return self.Z

    def setZ(self, z):
        self.Z = z

    def getZinv(self):
        return self.ZInv

    def setZInv(self, zinv):
        # FIXME Private members in OpenDSS
        self.ZInv = zinv

    def getKVBase(self):
        return self.kVBase

    def setKVBase(self, base):
        self.kVBase = base

    def getVMag(self):
        return self.VMag

    def setVMag(self, mag):
        self.VMag = mag

    def getPerUnit(self):
        return self.perUnit

    def setPerUnit(self, pu):
        self.perUnit = pu

    def getAngle(self):
        return self.angle

    def setAngle(self, ang):
        self.angle = ang

    def getEquivFrequency(self):
        return self.equivFrequency

    def setEquivFrequency(self, frequency):
        self.equivFrequency = frequency

    def getR1(self):
        return self.R1

    def setR1(self, r1):
        self.R1 = r1

    def getX1(self):
        return self.X1

    def setX1(self, x1):
        self.X1 = x1

    def getR0(self):
        return self.R0

    def setR0(self, r0):
        self.R0 = r0

    def getX0(self):
        return self.X0

    def setX0(self, x0):
        self.X0 = x0

    def isNeedToDoRecalc(self):
        return self.needToDoRecalc

    def setNeedToDoRecalc(self, value):
        self.needToDoRecalc = value

File: core/dss-core/src/main/java/com/ncond/dss/conversion/impl/EquivalentObjImpl.java
package com.ncond.dss.conversion.impl;

import java.io.PrintStream;

import com.ncond.dss.common.DSSClass;
import com.ncond.dss.common.SolutionObj;
import com.ncond.dss.common.impl.DSSClassImpl;
import com.ncond.dss.common.impl.DSSGlobals;
import com.ncond.dss.common.impl.Utilities;
import com.ncond.dss.conversion.Equivalent;
import com.ncond.dss.conversion.EquivalentObj;
import com.ncond.dss.parser.impl.Parser;
import com.ncond.dss.shared.CMatrix;
import com.ncond.dss.shared.impl.CMatrixImpl;
import com.ncond.dss.shared.impl.ComplexUtil;

import org.apache.commons.math.complex.Complex;

public class EquivalentObjImpl extends PCElementImpl implements EquivalentObj {

	private double kVBase;

	private double VMag;
	private double perUnit;
	private double angle;
	private double equivFrequency;

	private double[] R1, X1, R0, X0;

	private boolean needToDoRecalc;

	// base frequency series Z matrix
	protected CMatrix Z;
	protected CMatrix ZInv;

	public EquivalentObjImpl(DSSClassImpl parClass, String sourceName) {
		super(parClass);
		setName(sourceName.toLowerCase());
		objType = parClass.getDSSClassType(); //SOURCE + NON_PCPD_ELEM;  // don't want this in PC element list

		setNPhases(3);
		nConds = 3;
		setNTerms(1);
		Z    = null;
		ZInv = null;
		/*Basefrequency = 60.0;*/  // set in base class

		R1 = null;
		X1 = null;
		R0 = null;
		X0 = null;

		reallocRX();

		R1[0] = 1.65;
		X1[0] = 6.6;
		R0[0] = 1.9;
		X0[0] = 5.7;

		kVBase = 115.0;
		perUnit = 1.0;
		equivFrequency = baseFrequency;
		angle = 0.0;

		spectrum = "defaultvsource";

		initPropertyValues(0);

		YOrder = nTerms * nConds;
		recalcElementData();
	}

	private int idx(int a, int b) {
		return b * nTerms + a;
	}

	@Override
	public void recalcElementData() {
		Complex Zs, Zm;
//		int i, j, ii, jj;
		int iOffset, jOffset, indx;

		// for a source, nPhases = nCond, for now
		Z    = new CMatrixImpl(nPhases * nTerms);
		ZInv = new CMatrixImpl(nPhases * nTerms);

		// build Z matrix for all phases
		for (int i = 0; i < nTerms; i++)
			for (int j = 0; j < nTerms; j++) {
				indx = idx(i, j);
				Zs = ComplexUtil.divide(new Complex(2.0 * R1[indx] + R0[indx], 2.0 * X1[indx] + X0[indx] ), 3.0);
				Zm = ComplexUtil.divide(new Complex(R0[indx] - R1[indx], X0[indx] - X1[indx]), 3.0);

				iOffset = i * nPhases;
				jOffset = j * nPhases;

				for (int ii = 0; ii < nPhases; ii++) {
					for (int jj = 0; jj < ii; jj++) {
						if (ii == jj) {
							Z.set(ii + iOffset, jj + jOffset, Zs);
						} else {
							Z.set(ii + iOffset, jj + jOffset, Zm);
							Z.set(jj + iOffset, ii + jOffset, Zm);  // set other off-diagonal in this submatrix
						}
					}
				}

			}

		//  voltage source properties
		switch (nPhases) {
		case 1:
			VMag = kVBase * perUnit * 1000.0;
			break;
		default:
			VMag = kVBase * perUnit * 1000.0 / 2.0 / Math.sin((180.0 / nPhases)* Math.PI / 180.0);
			break;
		}

		setSpectrumObj((com.ncond.dss.general.SpectrumObj) DSSGlobals.spectrumClass.find(getSpectrum()));
		if (getSpectrumObj() == null)
			DSSGlobals.doSimpleMsg("Spectrum object \"" + getSpectrum() + "\" for device equivalent."+getName()+" not found.", 802);

		setInjCurrent( Utilities.resizeArray(getInjCurrent(), YOrder) );

		needToDoRecalc = false;
	}

	@Override
	public void calcYPrim() {
		Complex value;
		int i, j;
		double freqMultiplier;

		// build only YPrim series
		if (isYprimInvalid()) {
			YPrimSeries = new CMatrixImpl(YOrder);
			YPrim = new CMatrixImpl(YOrder);
		} else {
			YPrimSeries.clear();
			YPrim.clear();
		}

		if (needToDoRecalc)
			recalcElementData();

		YPrimFreq = DSSGlobals.activeCircuit.getSolution().getFrequency();
		freqMultiplier = YPrimFreq / baseFrequency;

		/* Put in Series RL matrix adjusted for frequency */
		for (i = 0; i < YOrder; i++) {
			for (j = 0; j < YOrder; j++) {
				value = Z.get(i, j);
				/* Modify from base freq */
				value = new Complex(value.getReal(), value.getImaginary() * freqMultiplier);
				ZInv.set(i, j, value);
			}
		}

		ZInv.invert();  /* Invert in place */

		if (ZInv.getErrorCode() > 0) {
			/* If error, put in large series conductance */
			DSSGlobals.doErrorMsg("EquivalentObj.calcYPrim", "Matrix inversion error for equivalent \"" + getName() + "\"",
					"Invalid impedance specified. Replaced with small resistance.", 803);
			ZInv.clear();
			for (i = 0; i < nPhases; i++)
				ZInv.set(i, i, new Complex(1.0 / DSSGlobals.EPSILON, 0.0));
		}

		YPrimSeries.copyFrom(ZInv);

		YPrim.copyFrom(YPrimSeries);

		/* Now account for open conductors */
		/* For any conductor that is open, zero out row and column */
		super.calcYPrim();

		setYPrimInvalid(false);
	}

	private void getVterminalForSource() {
		int i;
		Complex VHarm;
		double equivHarm;
		SolutionObj sol;

		try {
			/* This formulation will theoretically handle voltage sources of any number
			 * of phases assuming they are equally displaced in time.
			 */

			switch (nPhases) {
			case 1:
				VMag = kVBase * perUnit * 1000.0;
				break;
			default:
				VMag = kVBase * perUnit * 1000.0 / 2.0 / Math.sin((180.0 / nPhases) * Math.PI / 180.0);
				break;
			}

			sol = DSSGlobals.activeCircuit.getSolution();

			if (sol.isHarmonicModel()) {
				equivHarm = sol.getFrequency() / equivFrequency;
				VHarm = getSpectrumObj().getMult(equivHarm).multiply(VMag);  // base voltage for this harmonic
				VHarm = Utilities.rotatePhasorDeg(VHarm, equivHarm, angle);  // rotate for phase 1 shift
				for (i = 0; i < nPhases; i++) {
					VTerminal[i] = VHarm;
					if (i < nPhases)
						VHarm = Utilities.rotatePhasorDeg(VHarm, equivHarm, -360.0 / nPhases);
				}
			} else {
				for (i = 0; i < nPhases; i++)
					VTerminal[i] = ComplexUtil.polarDeg2Complex(VMag, (360.0 + angle - i * 360.0 / nPhases));
			}

		} catch (Exception e) {
			DSSGlobals.doSimpleMsg("Error computing voltages for Equivalent."+getName()+". Check specification. Aborting.", 804);
			if (DSSGlobals.inRedirect)
				DSSGlobals.redirectAbort = true;
		}
	}

	@Override
	public int injCurrents() {
		getInjCurrents(getInjCurrent());
		return super.injCurrents();  // add into system array
	}

	@Override
	public void getCurrents(Complex[] curr) {
		int i;
		SolutionObj sol;

		try {
			sol = DSSGlobals.activeCircuit.getSolution();
			for (i = 0; i < YOrder; i++)
				VTerminal[i] = sol.getNodeV(nodeRef[i]);

			YPrim.vMult(curr, VTerminal);

			getInjCurrents(complexBuffer);  // get present value of inj currents
			// add together with Yprim currents
			for (i = 0; i < YOrder; i++)
				curr[i] = curr[i].subtract(complexBuffer[i]);

		} catch (Exception e) {
			DSSGlobals.doErrorMsg(("GetCurrents for element: " + getName() + "."), e.getMessage(),
					"Inadequate storage allotted for circuit element.", 805);
		}
	}

	@Override
	public void getInjCurrents(Complex[] curr) {

		getVterminalForSource();
		YPrim.vMult(curr, VTerminal);  /* I = Y V */

		setITerminalUpdated(false);
	}

	@Override
	public void dumpProperties(PrintStream f, boolean complete) {
		int i, j;
		Complex c;

		super.dumpProperties(f, complete);

		DSSClass pc = getParentClass();
		for (i = 0; i < pc.getNumProperties(); i++)
			f.println("~ " + pc.getPropertyName()[i] + "=" + getPropertyValue(i));

		if (complete) {
			f.println();
			f.println("baseFrequency=" + baseFrequency);
			f.println("vMag=" + VMag);
			f.println("zMatrix=");
			for (i = 0; i < nPhases; i++) {
				for (j = 0; j < i; j++) {
					c = Z.get(i, j);
					f.print(c.getReal() + " + j" + c.getImaginary());
				}
				f.println();
			}
		}
	}

	@Override
	public void initPropertyValues(int arrayOffset) {

		setPropertyValue(0, "1");
		setPropertyValue(1, getBus(0));
		setPropertyValue(2, "115");
		setPropertyValue(3, "1");
		setPropertyValue(4, "0");
		setPropertyValue(5, "60");
		setPropertyValue(6, "3");
		setPropertyValue(7, "1.65");
		setPropertyValue(8, "6.6");
		setPropertyValue(9, "1.9");
		setPropertyValue(10, "5.7");

		super.initPropertyValues(Equivalent.NumPropsThisClass - 1);
	}

	@Override
	public String getPropertyValue(int index) {
		switch (index) {
		case 0:
			return null;
		default:
			return super.getPropertyValue(index);
		}
	}

	/**
	 * Make a positive sequence model.
	 */
	@Override
	public void makePosSequence() {
		String s;

		s = "Phases=1 ";
		s = s + String.format("basekV=%-.5g ", kVBase / DSSGlobals.SQRT3);
		s = s + String.format("r1=%-.5g ", R1);
		s = s + String.format("x1=%-.5g ", X1);

		Parser.getInstance().setCmdString(s);
		edit();

		super.makePosSequence();
	}

	// Private method in OpenDSS
	public int doTerminalsDef(int n) {
		int result = nTerms;

		if (n != nTerms)
			if (n > 0)
				reallocRX();

		return result;
	}

	/**
	 * Parse input string as an array.
	 */
	// Private method in OpenDSS
	public void parseDblMatrix(double[] mat) {
		Parser.getInstance().parseAsSymMatrix(nTerms, mat);
	}

	private void reallocRX() {
		R1 = Utilities.resizeArray(R1, (int) Math.pow(nTerms, 2));
		X1 = Utilities.resizeArray(X1, (int) Math.pow(nTerms, 2));
		R0 = Utilities.resizeArray(R0, (int) Math.pow(nTerms, 2));
		X0 = Utilities.resizeArray(X0, (int) Math.pow(nTerms, 2));
	}

	public CMatrix getZ() {
		return Z;
	}

	public void setZ(CMatrix z) {
		Z = z;
	}

	public CMatrix getZinv() {
		return ZInv;
	}

	public void setZInv(CMatrix zinv) {
		ZInv = zinv;
	}

	// FIXME Private members in OpenDSS

	public double getKvBase() {
		return kVBase;
	}

	public void setKvBase(double base) {
		this.kVBase = base;
	}

	public double getVMag() {
		return VMag;
	}

	public void setVMag(double mag) {
		VMag = mag;
	}

	public double getPerUnit() {
		return perUnit;
	}

	public void setPerUnit(double pu) {
		perUnit = pu;
	}

	public double getAngle() {
		return angle;
	}

	public void setAngle(double ang) {
		angle = ang;
	}

	public double getEquivFrequency() {
		return equivFrequency;
	}

	public void setEquivFrequency(double frequency) {
		equivFrequency = frequency;
	}

	public double[] getR1() {
		return R1;
	}

	public void setR1(double[] r1) {
		R1 = r1;
	}

	public double[] getX1() {
		return X1;
	}

	public void setX1(double[] x1) {
		X1 = x1;
	}

	public double[] getR0() {
		return R0;
	}

	public void setR0(double[] r0) {
		R0 = r0;
	}

	public double[] getX0() {
		return X0;
	}

	public void setX0(double[] x0) {
		X0 = x0;
	}

	public boolean isNeedToDoRecalc() {
		return needToDoRecalc;
	}

	public void setNeedToDoRecalc(boolean value) {
		needToDoRecalc = value;
	}

}
